### ISSUE TITLE
Random Network Distillation (RND) 적용을 통한 포머맨 학습

### ISSUE THREAD

**Issue by psyoblade** [modulabs-ctrl/pommerman-2018#4]
[Reinforcement Learning with Prediction-Based Rewards](https://blog.openai.com/reinforcement-learning-with-prediction-based-rewards/) 읽고 적용해보기
[정대님 pytorch 통한 RND 구현 github](https://github.com/kekmodel/rl_pytorch/blob/master/jupyter/rand_net_distill/dueling_double_dqn_with_rnd.ipynb)

**Comment by psyoblade** [comment]
[Large Scale Study of Curiosity-Driven Learning](https://pathak22.github.io/large-scale-curiosity/) 논문을 읽다보니 내용이 실험 환경에 대한 것이었음... 읽다가 [Curiosity-driven Exploration by Self-supervised Prediction](https://pathak22.github.io/noreward-rl/) 논문으로 넘어가기로 함. ;ㅁ;

**Comment by kekmodel** [comment]
좋습니다! 제 코드도 디버깅 중입니다아 ㅎㅎ

**Comment by psyoblade** [comment]
> 좋습니다! 제 코드도 디버깅 중입니다아 ㅎㅎ
최근에 논문이라는 것을 읽은 지가 꽤 된 것 같은데 ... 읽다보니 잘 쓰여진 논문을 읽는 것이 나름 즐겁구나 하는 생각도 들고 강화학습에서 기본적인 것들을 익힐 수도 있겠구나 싶었습니다. 다 읽고 한 번 공유해보려고 합니다.

**Comment by kekmodel** [comment]
RND의 핵심아이디어는 환경의 리워드에 호기심 리워드 보너스를 결합하는 것인데요.
호기심 리워드를 2개의 네트워크로 만듭니다.
- 하나는 랜덤초기화된 네트워크. (랜덤넷)
- 다른 하나는 랜덤넷 아랫단에 레이어를 몇개 추가한 네트워크 (예측넷)
액션의 결과인 s_t+1 을 각각의 네트워크에 인풋으로 넣어서 피쳐를 뽑습니다.
- s_t+1은 주기적으로 평균과 표준편차를 구해서 표준화 시켜줘서 인풋으로줍니다.
- loss function = MSE(랜덤피쳐, 예측피쳐)
호기심 리워드 = (랜덤넷피쳐 - 피쳐넷)^2
최종리워드 = (2*환경 리워드 + 호기심 리워드)/2
이렇게 하면 s_t+1에 의존적이면서 랜덤한 값(랜덤넷 피쳐)을 사용하여 탐험을 하는 것이고,
s_t+1의 변화가 크지 않다면 예측이 잘 되면서 그 값이 줄게 되어 호기심 보너스가 줄어듭니다.

**Comment by kekmodel** [comment]
멀티에이전트 알고리즘 중 nash Q 러닝이 라는 것이 있더군요. 시간될 때 적용해보려고 합니다.
http://downloads.hindawi.com/journals/complexity/2018/7172614.pdf

**Comment by psyoblade** [comment]
@kekmodel 님, 일전에 공유해 주신 코드를 참고해 https://github.com/modulabs-ctrl/pommerman-2018/blob/master/ctrl/cli/train_ppo_with_rnd.py#L234 구현을 적용해 보았는데요 문제는 pred_net 상에서 normalized 된 obs 값이 전부 nan 으로 나와서 f_loss 값들이 모두 nan 으로 나와 curiosity reward 값이 항상 0이 나오는 문제가 있습니다. 입력 observation 값들이 좀 치우친 sparse matrix 이긴 하지만 심하지는 않은데 pred_net 를 통과한 값들이 너무 작아서 그런걸까요?

**Comment by kekmodel** [comment]
현재 디버깅을 좀 해서 업데이트 했습니다.

**Comment by kekmodel** [comment]
그 문제 분모에 작은 엡실론을 클리핑해주는 방법으로 디버깅했습니다!

**Comment by psyoblade** [comment]
> 그 문제는 분모에 작은 엡실론을 클리핑해주는 방법으로 디버깅했습니다!
> 그리고 모멘트 값 계산 전에 step을 충분히 많이 밟아줘야 하더라구요!
> 
> ```
> EPS = np.finfo(np.float).eps
> def get_norm_params(obs_memory):
>     global obs_apace
> 
>     obses = [[] for _ in range(obs_space)]
>     for obs in obs_memory:
>         for j in range(obs_space):
>             obses[j].append(obs[j])
> 
>     mean = np.zeros(obs_space, 'float')
>     std = np.zeros(obs_space, 'float')
>     for i, obs_ in enumerate(obses):
>         mean[i] = np.mean(obs_)
>         std[i] = np.std(obs_)
>     obs_memory.clear()
> 
>     return mean, np.clip(std, a_min=EPS, a_max=None)
> ```

정대님 짱짱맨 ~ @kekmodel, 정말 감사합니다. clipping 값을 조금 조정해 주었고 intrinsic reward 값을 미조정해서 학습 해보고 있습니다. 제대로 학습 되는 지 여부는 좀 더 돌려봐야 알 수 있을 것 같습니다. :-)

**Comment by kekmodel** [comment]
리워드만 추가해서는 잘 작동 안할 수도 있습니다! PPO 내부구현을 바꿔보는 것도 좋을 듯 싶어요. 아니면 DQN으로 바꿔서 해볼까요?

**Comment by kekmodel** [comment]
일단은 스크립트 애들이랑 학습을 해서 그 친구들 보다 잘하면 우리 에이전트 끼리 학습시키는 것도 생각해볼만 한 것 같습니다. 너무 앞서가나요 ㅋㅋㅋ

**Comment by psyoblade** [comment]
> 일단은 스크립트 애들이랑 학습을 해서 그 친구들 보다 잘하면 우리 에이전트 끼리 학습시키는 것도 생각해볼만 한 것 같습니다. 너무 앞서가나요 ㅋㅋㅋ

의외로 이러한 접근이 더 유효할 수도 있을 것 같아요~ :-)

**Comment by psyoblade** [comment]
> 리워드만 추가해서는 잘 작동 안할 수도 있습니다! PPO 내부구현을 바꿔보는 것도 좋을 듯 싶어요. 아니면 DQN으로 바꿔서 해볼까요?
